File: src/com/colacelli/librebot/Librebot.java
package com.colacelli.librebot;

import com.colacelli.ircbot.IRCBot;
import com.colacelli.ircbot.plugins.access.AccessPlugin;
import com.colacelli.ircbot.plugins.apertiumtranslate.ApertiumTranslatePlugin;
import com.colacelli.ircbot.plugins.autojoin.AutoJoinPlugin;
import com.colacelli.ircbot.plugins.autoreconnect.AutoReconnectPlugin;
import com.colacelli.ircbot.plugins.ctcpversion.CTCPVersionPlugin;
import com.colacelli.ircbot.plugins.duckduckgosearch.DuckDuckGoSearchPlugin;
import com.colacelli.ircbot.plugins.help.HelpPlugin;
import com.colacelli.ircbot.plugins.ircop.IRCopPlugin;
import com.colacelli.ircbot.plugins.joinpart.JoinPartPlugin;
import com.colacelli.ircbot.plugins.loader.LoaderPlugin;
import com.colacelli.ircbot.plugins.nickserv.NickServPlugin;
import com.colacelli.ircbot.plugins.operator.OperatorPlugin;
import com.colacelli.ircbot.plugins.rejoinonkick.RejoinOnKickPlugin;
import com.colacelli.ircbot.plugins.rssfeed.RssFeedPlugin;
import com.colacelli.ircbot.plugins.thepiratebaysearch.ThePirateBaySearchPlugin;
import com.colacelli.ircbot.plugins.uptime.UptimePlugin;
import com.colacelli.ircbot.plugins.websitetitle.WebsiteTitlePlugin;
import com.colacelli.irclib.actors.Channel;
import com.colacelli.irclib.actors.User;
import com.colacelli.irclib.connection.Server;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Properties;

public class Librebot {
    private static final String PROPERTIES_SERVER = "server";
    private static final String PROPERTIES_PORT = "port";
    private static final String PROPERTIES_SSL = "ssl";
    private static final String PROPERTIES_PASSWORD = "password";
    private static final String PROPERTIES_IDENT = "ident";
    private static final String PROPERTIES_REAL_NAME = "real_name";
    private static final String PROPERTIES_NICK = "nick";
    private static final String PROPERTIES_CHANNELS = "channels";
    private static final String PROPERTIES_NICKSERV_PASSWORD = "nickserv_password";
    private static final String PROPERTIES_IRCOP_NAME = "ircop_name";
    private static final String PROPERTIES_IRCOP_PASSWORD = "ircop_password";
    private static final String PROPERTIES_CTCP_VERSION = "ctcp_version";
    private static final String PROPERTIES_FILE = "librebot.properties";


    public static void main(String[] args) {
        Properties properties = loadProperties();

        IRCBot bot = new IRCBot();

        addPlugins(bot, properties);

        bot.connect(
                buildServer(properties),
                buildUser(properties)
        );
    }

    private static Properties loadProperties() {
        Properties properties = new Properties();

        try {
            FileInputStream fileInputStream = new FileInputStream(PROPERTIES_FILE);
            properties.load(fileInputStream);
        } catch (IOException e) {
            e.printStackTrace();
        }

        return properties;
    }

    private static User buildUser(Properties properties) {
        User.Builder builder = new User.Builder();
        builder
                .setNick(properties.getProperty(PROPERTIES_NICK))
                .setLogin(properties.getProperty(PROPERTIES_IDENT))
                .setRealName(properties.getProperty(PROPERTIES_REAL_NAME));

        return builder.build();
    }

    private static Server buildServer(Properties properties) {
        Server.Builder builder = new Server.Builder();
        builder
                .setHostname(properties.getProperty(PROPERTIES_SERVER))
                .setPort(Integer.parseInt(properties.getProperty(PROPERTIES_PORT)))
                .setSecure(Boolean.parseBoolean(properties.getProperty(PROPERTIES_SSL)))
                .setPassword(properties.getProperty(PROPERTIES_PASSWORD));

        return builder.build();
    }

    private static void addPlugins(IRCBot bot, Properties properties) {
        ArrayList<Channel> channels = new ArrayList<>();
        for (String channel : properties.getProperty(PROPERTIES_CHANNELS).split(",")) {
            channels.add(new Channel(channel));
        }

        bot.addPlugin(new AccessPlugin());
        bot.addPlugin(new ApertiumTranslatePlugin());
        bot.addPlugin(new AutoJoinPlugin(channels));
        bot.addPlugin(new AutoReconnectPlugin());
        bot.addPlugin(new CTCPVersionPlugin(properties.getProperty(PROPERTIES_CTCP_VERSION)));
        bot.addPlugin(new DuckDuckGoSearchPlugin());
        bot.addPlugin(new HelpPlugin());
        bot.addPlugin(new JoinPartPlugin());
        bot.addPlugin(new LoaderPlugin());
        bot.addPlugin(new OperatorPlugin());
        bot.addPlugin(new RejoinOnKickPlugin());
        bot.addPlugin(new RssFeedPlugin());
        bot.addPlugin(new ThePirateBaySearchPlugin());
        bot.addPlugin(new UptimePlugin());
        bot.addPlugin(new WebsiteTitlePlugin());

        String nickservPassword = properties.getProperty(PROPERTIES_NICKSERV_PASSWORD);
        if (!nickservPassword.isEmpty()) bot.addPlugin(new NickServPlugin(nickservPassword));

        String ircopName = properties.getProperty(PROPERTIES_IRCOP_NAME);
        String ircopPassword = properties.getProperty(PROPERTIES_IRCOP_PASSWORD);
        if (!ircopName.isEmpty() && !ircopPassword.isEmpty()) bot.addPlugin(new IRCopPlugin(ircopName, ircopPassword));
    }
}

File: src/com/colacelli/librebot/Librebot.kt
package com.colacelli.librebot

import com.colacelli.ircbot.IRCBot
import com.colacelli.ircbot.plugins.access.AccessPlugin
import com.colacelli.ircbot.plugins.apertiumtranslate.ApertiumTranslatePlugin
import com.colacelli.ircbot.plugins.autojoin.AutoJoinPlugin
import com.colacelli.ircbot.plugins.autoreconnect.AutoReconnectPlugin
import com.colacelli.ircbot.plugins.autoresponse.AutoResponsePlugin
import com.colacelli.ircbot.plugins.ctcpversion.CTCPVersionPlugin
import com.colacelli.ircbot.plugins.duckduckgosearch.DuckDuckGoSearchPlugin
import com.colacelli.ircbot.plugins.help.HelpPlugin
import com.colacelli.ircbot.plugins.ircop.IRCopPlugin
import com.colacelli.ircbot.plugins.joinpart.JoinPartPlugin
import com.colacelli.ircbot.plugins.nickserv.NickServPlugin
import com.colacelli.ircbot.plugins.operator.OperatorPlugin
import com.colacelli.ircbot.plugins.pluginloader.PluginLoaderPlugin
import com.colacelli.ircbot.plugins.rejoinonkick.RejoinOnKickPlugin
import com.colacelli.ircbot.plugins.rssfeed.RSSFeedPlugin
import com.colacelli.ircbot.plugins.thepiratebaysearch.ThePirateBaySearchPlugin
import com.colacelli.ircbot.plugins.uptime.UptimePlugin
import com.colacelli.ircbot.plugins.websitetitle.WebsiteTitlePlugin
import com.colacelli.irclib.actors.Channel
import com.colacelli.irclib.actors.User
import com.colacelli.irclib.connection.Server
import java.io.FileInputStream
import java.util.*
import kotlin.collections.ArrayList

class Librebot {
    companion object {
        private const val PROPERTIES_HOSTNAME = "hostname"
        private const val PROPERTIES_PORT = "port"
        private const val PROPERTIES_SSL = "ssl"
        private const val PROPERTIES_PASSWORD = "password"
        private const val PROPERTIES_LOGIN = "login"
        private const val PROPERTIES_REAL_NAME = "real_name"
        private const val PROPERTIES_NICK = "nick"
        private const val PROPERTIES_CHANNELS = "channels"
        private const val PROPERTIES_NICKSERV_PASSWORD = "nickserv_password"
        private const val PROPERTIES_IRCOP_NAME = "ircop_name"
        private const val PROPERTIES_IRCOP_PASSWORD = "ircop_password"
        private const val PROPERTIES_CTCP_VERSION = "ctcp_version"
        private const val PROPERTIES_FILE = "librebot.properties"

        @JvmStatic
        fun main(args: Array<String>) {

            val properties = Properties()
            properties.load(FileInputStream(PROPERTIES_FILE))

            val server = Server(
                    properties.getProperty(PROPERTIES_HOSTNAME, "irc.freenode.net"),
                    properties.getProperty(PROPERTIES_PORT, "6697").toInt(),
                    properties.getProperty(PROPERTIES_SSL, "true").toBoolean(),
                    properties.getProperty(PROPERTIES_PASSWORD, "")
            )

            val user = User(
                    properties.getProperty(PROPERTIES_NICK, "librebot"),
                    properties.getProperty(PROPERTIES_LOGIN, "librebot"),
                    properties.getProperty(PROPERTIES_REAL_NAME, "GNU Librebot - https://gitlab.com/jic/librebot")
            )

            val bot = IRCBot(server, user)

            val channels = ArrayList<Channel>()
            properties.getProperty(PROPERTIES_CHANNELS, "#gnu").split(",").forEach {
                channels.add(Channel(it))
            }

            bot.addPlugin(AccessPlugin())
            bot.addPlugin(ApertiumTranslatePlugin())
            bot.addPlugin(AutoJoinPlugin(channels))
            bot.addPlugin(AutoReconnectPlugin())
            bot.addPlugin(AutoResponsePlugin())
            bot.addPlugin(CTCPVersionPlugin(properties.getProperty(PROPERTIES_CTCP_VERSION, "GNU Librebot - https://gitlab.com/jic/librebot")))
            bot.addPlugin(DuckDuckGoSearchPlugin())
            bot.addPlugin(HelpPlugin())
            bot.addPlugin(JoinPartPlugin())
            bot.addPlugin(PluginLoaderPlugin())
            bot.addPlugin(OperatorPlugin())
            bot.addPlugin(RejoinOnKickPlugin())
            bot.addPlugin(RSSFeedPlugin())
            bot.addPlugin(ThePirateBaySearchPlugin())
            bot.addPlugin(UptimePlugin())
            bot.addPlugin(WebsiteTitlePlugin())

            val nickservPassword = properties.getProperty(PROPERTIES_NICKSERV_PASSWORD, "")
            if (nickservPassword.isNotBlank()) bot.addPlugin(NickServPlugin(nickservPassword))

            val ircopName = properties.getProperty(PROPERTIES_IRCOP_NAME, "")
            val ircopPassword = properties.getProperty(PROPERTIES_IRCOP_PASSWORD, "")
            if (ircopName.isNotBlank() && ircopPassword.isNotBlank()) bot.addPlugin(IRCopPlugin(ircopName, ircopPassword))

            bot.connect()
        }
    }
}